src/metabase/db/liquibase.clj

```diff
@@ -261,6 +238,40 @@
                        (finally
                          (.releaseLock lock-service))))))))
 
+(defmacro with-scope-locked
+  "Run `body` in a scope on the Liquibase instance `liquibase`.
+   Liquibase scopes are used to hold configuration and parameters (akin to binding dynamic variables in
+   Clojure). This function initializes the database and the resource accessor which are often required.
+   The underlying locks are re-entrant, so it is safe to nest these blocks."
+  {:style/indent 1}
+  [liquibase & body]
+  `(run-in-scope-locked ~liquibase (fn [] ~@body)))
+
+(defn migrate-up-if-needed!
+  "Run any unrun `liquibase` migrations, if needed."
+  [^Liquibase liquibase ^DataSource data-source]
+  (log/info (trs "Checking if Database has unrun migrations..."))
+  (if (seq (unrun-migrations data-source))
+    (do
+     (log/info (trs "Database has unrun migrations. Checking if migration lock is taken..."))
+     (wait-for-migration-lock-to-be-cleared liquibase)
```

Comment:
hmmm, this might no longer be needed. `with-scope-locked` will wait for locks anyway, what do you think if we bake this behavior into `with-scope-locked`?

Reply:
Yes I agree it feels awkward having both methods. Without some refactoring, it's best to leave the redundancy, as `wait-for-migration-lock-to-be-cleared` provides a better error message that mentions the `release-locks` command.

Reply:
I've consolidated the two methods. This fixes a subtle case where the lock could have been taken again between waiting and taking it ourselves, which would have resulted in a very long additional wait, and a totally different lock exception being thrown by Liquibase itself.